### PR TITLE
999 Sidebar filters and datasets

### PR DESCRIPTION
**Pull request by jorgeatgu**
Closes https://github.com/PopulateTools/issues/issues/999

## :v: What does this PR do?

- Select any category or frequency, the dataset lists will be filtered and updated.

- If we select a dataset, it's placed at the top of the list.

## :mag: How should this be manually tested?

Staging

### Before this PR

999_before 2020-04-15 16_18_45

### After this PR

999_after 2020-04-15 16_28_03

### CODE REVIEW

**Comment by furilo** [comment]
I'd add in this PR that if we enter into a dataset, the sidebar changes to the second tab.

**Comment by ferblape** [comment]
Also, scroll to top when clicking a dataset:

scroll-data

File: app/javascript/gobierto_data/webapp/components/sidebar/SidebarCategories.vue [hunk 1]
```diff
@@ -75,10 +75,12 @@ export default {
     this.labelCategories = I18n.t("gobierto_data.projects.categories")
   },
   methods: {
-    sendCheckboxStatus({ id, value, filter }) {
+    sendCheckboxStatus_TEMP({ id, value, filter }) {
```
Comment (ferblape) [inline app/javascript/gobierto_data/webapp/components/sidebar/SidebarCategories.vue:88]: Apart from the _TEMP name add a TODO or FIXME comment, please

File: app/javascript/gobierto_data/webapp/pages/Main.vue [hunk 2]
```diff
@@ -37,5 +37,13 @@ export default {
       default: 0
     }
   },
+  created() {
+    this.$root.$on("sendCheckbox", this.handleCheckboxStatus)
+    this.$root.$on("selectAll", this.handleIsEverythingChecked)
```
Comment (Crashillo) [inline app/javascript/gobierto_data/webapp/pages/Main.vue]: Rename these event names also with _TEMP

File: app/javascript/gobierto_data/webapp/components/sidebar/SidebarDatasets.vue [hunk 3]
```diff
@@ -98,13 +99,13 @@ export default {
...
-      const indexToggle = allDatasets.findIndex(dataset => dataset.attributes.slug == slug)
+      const indexToggle = allDatasets.findIndex(dataset => dataset.attributes.slug === slug)
       this.toggle = indexToggle
       if (this.toggle === -1) {
         this.toggle = 0
         slug = slug = allDatasets.length ? allDatasets[0].attributes.slug : ''
       }
-      let firstElement = allDatasets.find(dataset => dataset.attributes.slug == slug)
+      let firstElement = allDatasets.find(dataset => dataset.attributes.slug === slug)
       let filteredArray = allDatasets.filter(dataset => dataset.attributes.slug !== slug)
```
Comment (Crashillo) [inline app/javascript/gobierto_data/webapp/components/sidebar/SidebarDatasets.vue]: Apply destructuring all over this file. Example:

```
let { id } = this.$route.params || {}
const indexToggle = allDatasets.findIndex(({ attributes: { slug } = {}}) => slug === id)
```

You can only trust on those objects from external libraries, i.e. vue-router exposes always a `params` property.

**Review by Crashillo — CHANGES_REQUESTED** [review]

File: app/javascript/gobierto_data/webapp/components/sidebar/SidebarDatasets.vue (hunk 3 above)
Comment (Crashillo) [reply app/javascript/gobierto_data/webapp/components/sidebar/SidebarDatasets.vue]: Do this kind: `allDatasets[0].attributes.slug` should be forbidden by law

**Comment by jorgeatgu** [comment]
- Include scroll to top when user click on dataset
- Change sidebar tab when user click on dataset from home
- Fixed error when access to dataset from link

File: app/javascript/gobierto_data/webapp/components/sidebar/SidebarCategories.vue [hunk 4]
```diff
@@ -5,43 +5,60 @@
...
-        :class="!filter.key ? 'gobierto-data-filters-element-no-margin' : ''"
+        :class="!filter.isToggle ? 'gobierto-data-filters-element-no-margin' : ''"
         class="gobierto-data-filters-element"
       >
-        <template v-if="filter.type === 'vocabulary_options'">
-          <div>
-            <BlockHeader
+        <Dropdown @is-content-visible="filter.isToggle = !filter.isToggle">
+          <template v-slot:trigger>
+            <!-- <BlockHeader
```
Comment (ferblape) [inline app/javascript/gobierto_data/webapp/components/sidebar/SidebarCategories.vue]: Remove this commented code

File: app/javascript/gobierto_data/webapp/components/sidebar/SidebarCategories.vue [hunk 5]
```diff
@@ -5,43 +5,60 @@
...
               :label-alt="filter.isEverythingChecked"
-              :class="filter.key ? '' : 'gobierto-filter-rotate-icon'"
+              :class="filter.isToggle ? '' : 'gobierto-filter-rotate-icon'"
               see-link
-              @toggle="e => filter.key = !filter.key"
-              @select-all="e => selectAllCheckbox({ ...e, filter })"
+              @select-all="e => selectAllCheckbox_TEMP({ ...e, filter })"
+            /> -->
+            <!-- FIXME: remove inline styles -->
+            <div class="gobierto-block-header"
```
[the remark this answers is not in the document]
Comment (ferblape) [reply app/javascript/gobierto_data/webapp/components/sidebar/SidebarCategories.vue]: Fix

File: app/javascript/gobierto_data/webapp/components/sidebar/SidebarCategories.vue [hunk 6]
```diff
@@ -5,43 +5,60 @@
...
+              style="flex: 1;"
```
[the remark this answers is not in the document]
Comment (ferblape) [reply app/javascript/gobierto_data/webapp/components/sidebar/SidebarCategories.vue:18]: Fix

**Review by ferblape — CHANGES_REQUESTED** [review]
We are almost there!

**Review by ferblape — APPROVED** [review]

File: app/javascript/gobierto_data/webapp/components/sidebar/SidebarDatasets.vue [hunk 7]
```diff
@@ -98,13 +99,13 @@ export default {
       const sortDatasets = this.items
       const allDatasets = sortDatasets.sort((a, b) => a.attributes.name.localeCompare(b.attributes.name));
       let slug = this.$route.params.id
-      const indexToggle = allDatasets.findIndex(dataset => dataset.attributes.slug == slug)
+      const indexToggle = allDatasets.findIndex(dataset => dataset.attributes.slug === slug)
       this.toggle = indexToggle
       if (this.toggle === -1) {
         this.toggle = 0
         slug = slug = allDatasets.length ? allDatasets[0].attributes.slug : ''
```
Comment (ferblape) [inline app/javascript/gobierto_data/webapp/components/sidebar/SidebarDatasets.vue]: what does `slug = slug = allDatasets`? I think one of the `slug =` is duplicated

File: app/javascript/gobierto_data/webapp/components/sidebar/SidebarDatasets.vue (hunk 3 above)
Comment (ferblape) [reply app/javascript/gobierto_data/webapp/components/sidebar/SidebarDatasets.vue]: @jorgeatgu this is the only comment left to merge the PR

File: app/javascript/gobierto_data/webapp/components/Sidebar.vue [hunk 8]
```diff
@@ -53,24 +54,31 @@ export default {
...
       labelQueries: I18n.t("gobierto_data.projects.queries") || "",
       labelCategories: I18n.t("gobierto_data.projects.categories") || "",
-      currentTabComponent: null,
-      currentTab: this.activeTab
+      currentTabComponent: null
     };
   },
+  watch: {
+    activeTab(newValue) {
+      this.currentTabComponent = COMPONENTS[newValue];
```
Comment (Crashillo) [inline app/javascript/gobierto_data/webapp/components/Sidebar.vue]: Wrap these line with conditionals
```suggestion
    activeTab(newValue, oldValue) {
    if (newValue !== oldValue) {
          this.currentTabComponent = COMPONENTS[newValue];
    }
```
Always do so when you watch properties, otherwise this could be an infinite loop

File: app/javascript/gobierto_data/webapp/components/sets/DataTab.vue [hunk 9]
```diff
@@ -42,10 +42,6 @@ export default {
       type: Array,
       required: true
     },
```
Comment (Crashillo) [inline app/javascript/gobierto_data/webapp/components/sets/DataTab.vue:44]: Why do you remove `publicQueries` as a prop? Why is defined as a `data`?

File: app/javascript/gobierto_data/webapp/components/sidebar/SidebarCategories.vue [hunk 10]
```diff
@@ -58,28 +68,37 @@ export default {
...
+    filtersModify() {
+      return this.filters.length ? this.filters.map(d => ({ ...d, isToggle: true })) : []
     }
   },
   created() {
     this.labelSets = I18n.t("gobierto_data.projects.sets")
     this.labelQueries = I18n.t("gobierto_data.projects.queries")
     this.labelCategories = I18n.t("gobierto_data.projects.categories")
+    this.labelAll = I18n.t("gobierto_common.vue_components.block_header.all");
+    this.labelNone = I18n.t("gobierto_common.vue_components.block_header.none");
```
Comment (Crashillo) [inline app/javascript/gobierto_data/webapp/components/sidebar/SidebarCategories.vue]: Move all these stuff inside `data()` like:

`labelSets: I18n.t("gobierto_data.projects.sets")  || "",`

File: app/javascript/gobierto_data/webapp/components/sidebar/SidebarDatasets.vue (hunk 7 above)
Comment (jorgeatgu) [reply app/javascript/gobierto_data/webapp/components/sidebar/SidebarDatasets.vue]: Fixed here: https://github.com/PopulateTools/gobierto/blob/d3bde171a5a52a31a5d0d6ad121161697abe0244/app/javascript/gobierto_data/webapp/components/sidebar/SidebarDatasets.vue#L117

File: app/javascript/gobierto_data/webapp/components/sidebar/SidebarDatasets.vue [hunk 11]
```diff
@@ -95,17 +107,17 @@ export default {
   },
   methods: {
     orderDatasets() {
-      const sortDatasets = this.items
-      const allDatasets = sortDatasets.sort((a, b) => a.attributes.name.localeCompare(b.attributes.name));
-      let slug = this.$route.params.id
-      const indexToggle = allDatasets.findIndex(dataset => dataset.attributes.slug == slug)
+      const allDatasets = this.items.sort((a, b) => a.attributes.name.localeCompare(b.attributes.name));
```
Comment (Crashillo) [inline app/javascript/gobierto_data/webapp/components/sidebar/SidebarDatasets.vue]: Same props destructuration as beneath

File: app/javascript/gobierto_data/webapp/components/sidebar/SidebarDatasets.vue [hunk 12]
```diff
@@ -95,17 +107,17 @@ export default {
...
+      let { id } = this.$route.params || {}
+
+      const indexToggle = allDatasets.findIndex(({ attributes: { slug } = {} }) => slug === id)
       this.toggle = indexToggle
       if (this.toggle === -1) {
         this.toggle = 0
-        slug = slug = allDatasets.length ? allDatasets[0].attributes.slug : ''
+        id = allDatasets.length ? allDatasets[0].attributes.slug : ''
```
Comment (Crashillo) [inline app/javascript/gobierto_data/webapp/components/sidebar/SidebarDatasets.vue]: You're mixing unrelated stuff. 

If you need the `id` from $route, you set it in a variable, but don't reassign it, why so? The second `id` is a totally different variable, then you create a new variable for it. These practices are the opposite of code readability.

Then, one more time, destructuration. Why attributes ought to have a property called `slug`?
```
let anotherId = null;
if (this.toggle.....) {
    ...other stuff

    // destructuration prop by prop, fallbacking them
    const [destructuratedVar: { attributes: { slug } = {} } = {} ] = allDatasets || []
    anotherId = slug

   // same in one line (more complex syntax)
   ([destructuratedVar: { attributes: { slug: anotherId } = {} } = {} ] = allDatasets || [])
}
```

File: app/javascript/gobierto_data/lib/router.js [hunk 13]
```diff
@@ -17,20 +17,24 @@ export const router = new VueRouter({
...
         path: ":id/q/:queryId?",
         component: Dataset,
-        // active the editor for queries
-        props: { activeDatasetTab: 1, activeSidebarTab: 1 }
+        name: 'Dataset',
+        props: { activeDatasetTab: 1 }
       }]
-    }]
+    }],
+    scrollBehavior() {
```
Comment (Crashillo) [inline app/javascript/gobierto_data/lib/router.js:34]: Why is this here?

**Review by Crashillo — CHANGES_REQUESTED** [review]

File: app/javascript/gobierto_data/webapp/components/sidebar/SidebarDatasets.vue (hunk 7 above)
Comment (Crashillo) [reply app/javascript/gobierto_data/webapp/components/sidebar/SidebarDatasets.vue]: Take into account the Github label `Outdated` ( ͡° ͜ʖ ͡°)

File: app/javascript/gobierto_data/lib/router.js (hunk 13 above)
Comment (jorgeatgu) [reply app/javascript/gobierto_data/lib/router.js:34]: Because I read it in the [documentation](https://router.vuejs.org/guide/advanced/scroll-behavior.html#async-scrolling), where should I be?

Comment (Crashillo) [reply app/javascript/gobierto_data/lib/router.js:34]: Aaaaaaamigo, it's a predefined method. Ok, then. Cool.

File: app/javascript/gobierto_data/webapp/components/sidebar/SidebarDatasets.vue [hunk 14]
```diff
@@ -95,17 +107,27 @@ export default {
   },
   methods: {
     orderDatasets() {
-      const sortDatasets = this.items
-      const allDatasets = sortDatasets.sort((a, b) => a.attributes.name.localeCompare(b.attributes.name));
-      let slug = this.$route.params.id
-      const indexToggle = allDatasets.findIndex(dataset => dataset.attributes.slug == slug)
+      //TODO Datasets should be order in filter mixin
+      const allDatasets = this.items.sort((a, b) => a.attributes.name.localeCompare(b.attributes.name));
```
Comment (Crashillo) [inline app/javascript/gobierto_data/webapp/components/sidebar/SidebarDatasets.vue]: Same props destructuration means:
```suggestion
      const allDatasets = this.items.sort(({ attributes: { name: a } = {} }, { attributes: { name: b } = {} }) => a.localeCompare(b));
```

**Review by Crashillo — CHANGES_REQUESTED** [review]

**Review by Crashillo — APPROVED** [review]

**Comment by ferblape** [comment]
I'm merging this now it's approved by Hugo, I don't want to accumulate PRs of gobierto data.
